### PR TITLE
[1096] - Add cookie banner

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/i0soP8KH/1096-m-add-cookie-banner

### Changes proposed in this pull request

- Implement the Gov UK cookie banner and wire in some Javascript to show/hide it. It's always shown without JS as per the design guide
- Add Jest and some unit tests and a rake task to run it in our CI
- Fix a few of the feature specs which were generically targeting `govuk-button` causing capybara ambiguous errors

### Guidance to review

- Head to the review app https://rtt-review-pr-589.herokuapp.com/
- Assert you can see the cookie banner. Select to hide it, refresh and assert it is no longer shown.
- Disable Javascript in the console and assert the banner is still shown but without the hide button

### CODE REVIEW

**Comment by edwardhorsford** [comment]
Looking good @darokel.

Note the Design system has guidance for how to hide banner - they [suggest using a `hidden` attribute, and not `display="none"`](https://design-system.service.gov.uk/components/cookie-banner/#when-the-user-accepts-or-rejects-cookies-using-the-cookie-banner).

We should also add the `viewed_cookie_message` cookie to our cookie page.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -65,9 +65,12 @@ jobs:
     - name: Run SCSS Linter
       run:  docker-compose exec -T web /bin/sh -c 'bundle exec rake lint:scss'
 
-    - name: Run tests
+    - name: Run Rspec tests
       run: docker-compose exec --env COVERAGE=true -T web /bin/sh -c 'bundle exec rake spec'
 
+    - name: Run Javascript tests
+      run: docker-compose exec --env -T web /bin/sh -c 'yarn run tests'
```
Comment (AbigailMcP) [inline .github/workflows/build.yml:72]: Does this need to be `yarn run test`?

File: Rakefile [hunk 2]
```diff
@@ -8,5 +8,11 @@ require_relative "config/application"
 Rails.application.load_tasks
 
 if %w[development test].include?(Rails.env)
-  task default: ["lint:ruby", "lint:javascript", "lint:scss", :spec]
+
+  desc "Runs JS unit tests with yarn"
+  task javascript_specs: :environment do
+    system "yarn test"
```
Comment (AbigailMcP) [inline Rakefile:14]: Do we need to add something to the README re installing jest globally on your local machine?

**Review by AbigailMcP — COMMENTED** [review]
Looks good! Just a couple of comments inline.

File: Rakefile (hunk 2 above)
Comment (darokel) [reply Rakefile:14]: It should be installed when you run yarn as we've declared it in the project as a dev dependency.

File: .github/workflows/build.yml (hunk 1 above)
Comment (darokel) [reply .github/workflows/build.yml]: yup good spot fixed in https://github.com/DFE-Digital/register-trainee-teachers/pull/589/commits/d0369e3d70b1b5c41ce280554dfbbe267c1e4125

File: app/views/sign_in/index.html.erb [hunk 3]
```diff
@@ -15,9 +15,9 @@
     <p class="govuk-body-m"> Use DfE Sign-in to access your account.</p>
 
     <% if FeatureService.enabled?("use_dfe_sign_in")%>
-      <%= govuk_button_to("Sign in using DfE Sign-in", "/auth/dfe") %>
+      <%= govuk_button_to("Sign in using DfE Sign-in", "/auth/dfe", class: "use_dfe_sign_in") %>
     <% else %>
-      <%= govuk_button_link_to "Sign in using Persona", "/personas" %>
+      <%= govuk_button_link_to "Sign in using Persona", "/personas", class: "use_dfe_sign_in" %>
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/views/sign_in/index.html.erb]: Discussed offline, used `qa-` as a prefix to be consistent with other repos

**Comment by darokel** [comment]
@edwardhorsford I've applied those changes

**Review by AbigailMcP — APPROVED** [review]

File: app/views/pages/cookies.html.erb [hunk 4]
```diff
@@ -3,12 +3,23 @@
...
+    
     <h3 class="govuk-heading-s">A cookie to give you access to your account</h3>
+    
     <p class="govuk-body">"Register_trainee_teachers_session" is a cookie that reminds the service who you are when you’re signed in.</p>
+    
     <p class="govuk-body">It enables you to access the information in your account, such as your trainee records.</p>
+    
     <p class="govuk-body">It expires when you sign out, or after 6 hours on the service.</p>
+
+    <h3 class="govuk-heading-s">A cookie to store that you have acknowledged our essential cookies banner.</h3>
```
Comment (edwardhorsford) [inline app/views/pages/cookies.html.erb:21]:
```suggestion
    <h3 class="govuk-heading-s">A cookie to store that you have seen the cookies banner.</h3>
```

File: app/views/pages/cookies.html.erb [hunk 5]
```diff
@@ -3,12 +3,23 @@
...
+
+    <p class="govuk-body">"viewed_cookie_message" is a cookie which tells the service that you have acknowledged our essential cookies banner so we do not show it to you again.</p>
```
Comment (edwardhorsford) [inline app/views/pages/cookies.html.erb:23]:
```suggestion
    <p class="govuk-body">"viewed_cookie_message" is a cookie which tells the service that you have the cookies banner so we do not show it to you again.</p>
```

**Review by edwardhorsford — APPROVED** [review]

File: app/views/pages/cookies.html.erb (hunk 5 above)
Comment (darokel) [reply app/views/pages/cookies.html.erb:23]:
```suggestion
<p class="govuk-body">"viewed_cookie_message" is a cookie which tells the service that you have seen the cookies banner so we do not show it to you again.</p>
```

**Pull request merged**